### PR TITLE
Query refactoring: SpanOrQueryBuilder and Parser

### PR DESCRIPTION
**Pull request by cbuescher**
Moving the query building functionality from the parser to the builders
new doToQuery() method analogous to other recent query refactorings.

Relates to #10217

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/index/query/SpanOrQueryBuilder.java [hunk 1]
```diff
@@ -19,24 +19,38 @@
...
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
 import org.elasticsearch.common.xcontent.XContentBuilder;
 
 import java.io.IOException;
 import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
 
 public class SpanOrQueryBuilder extends AbstractQueryBuilder<SpanOrQueryBuilder> implements SpanQueryBuilder<SpanOrQueryBuilder> {
```
Comment (MaineC) [inline core/src/main/java/org/elasticsearch/index/query/SpanOrQueryBuilder.java]: Nitpick: With this turning into a public API - should we have some class level doc here?

File: core/src/main/java/org/elasticsearch/index/query/SpanOrQueryParser.java [hunk 2]
```diff
@@ -34,7 +31,7 @@
 /**
  *
  */
```
Comment (MaineC) [inline core/src/main/java/org/elasticsearch/index/query/SpanOrQueryParser.java]: Either fill with real information or remove the empty comment.

File: core/src/main/java/org/elasticsearch/index/query/SpanOrQueryParser.java [hunk 3]
```diff
@@ -34,7 +31,7 @@
...
-public class SpanOrQueryParser extends BaseQueryParserTemp {
+public class SpanOrQueryParser extends BaseQueryParser {
 
     @Inject
```
Comment (MaineC) [inline core/src/main/java/org/elasticsearch/index/query/SpanOrQueryParser.java]: Shouldn't be needed for a constructor w/o arguments.

File: core/src/test/java/org/elasticsearch/index/query/SpanOrQueryBuilderTest.java [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+        }
+        return new SpanOrQuery(spanQueries);
+    }
+
+    @Override
+    protected SpanOrQueryBuilder doCreateTestQueryBuilder() {
+        SpanOrQueryBuilder queryBuilder = new SpanOrQueryBuilder();
+        int clauses = randomIntBetween(1, 6);
+        // we use one random SpanTermQueryBuilder to determine same field name for subsequent clauses
+        String fieldName = new SpanTermQueryBuilderTest().createTestQueryBuilder().fieldName();
```
Comment (MaineC) [inline core/src/test/java/org/elasticsearch/index/query/SpanOrQueryBuilderTest.java]: Is this something that could be moved to a more general utility class for testing span queries?

W/o having looked at the implementation: Or at least something that could be moved to a static helper method of the test class itself?

**Comment by MaineC** [comment]
Left some minor comments.

**Comment by cbuescher** [comment]
@MaineC thanks for the review, just pushed new commit addressing your comments.

**Comment by MaineC** [comment]
LGTM - I'm sure @javanna will will find a few more areas for improvment

File: core/src/main/java/org/elasticsearch/index/query/SpanOrQueryBuilder.java [hunk 5]
```diff
@@ -53,6 +70,55 @@ protected void doXContent(XContentBuilder builder, Params params) throws IOExcep
     }
 
     @Override
+    protected Query doToQuery(QueryParseContext parseContext) throws IOException {
+        SpanQuery[] spanQueries = new SpanQuery[clauses.size()];
+        for (int i = 0; i < clauses.size(); i++) {
+            Query query = clauses.get(i).toQuery(parseContext);
+            assert query instanceof SpanQuery;
```
Comment (javanna) [inline core/src/main/java/org/elasticsearch/index/query/SpanOrQueryBuilder.java]: we might have the same problem with LateParsingQuery here too right? then this should become an if instead? I guess we might have the same somewhere else too?

**Comment by javanna** [comment]
left one small comment

File: core/src/main/java/org/elasticsearch/index/query/SpanOrQueryBuilder.java (hunk 5 above)
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/index/query/SpanOrQueryBuilder.java]: Not really, all query clauses here need to be SpanQueryBuilders, the problem with LateParsingQuery was related to MultiTermQueryBuilders (which RangeQuery is a subtype of). We are creating SpanOrQuery here which takes list of SpanQueries as an argument, so all good I think.

Comment (javanna) [inline core/src/main/java/org/elasticsearch/index/query/SpanOrQueryBuilder.java]: ok sorry for the confusion, here and in the other PR where I commented...

File: core/src/test/java/org/elasticsearch/index/query/SpanOrQueryBuilderTest.java (hunk 4 above)
Comment (cbuescher) [inline core/src/test/java/org/elasticsearch/index/query/SpanOrQueryBuilderTest.java]: Not quiet sure which part you think is general enough to factor our here. I briefly tried having a general Span-Query producing utility method in RandomQueryBuilder but the problem here is that ony SpanTermQueryBuilder is a leaf query and for the others we would have to deal with the depth of nested subtrees. Just getting a list of span queries with the same fieldname is only needed here, the other span queries come either with one or two subclauses only.

**Comment by cbuescher** [comment]
Addressed comments and removed check for empty clauses in doXContent since this will either be caught by parser or in validate. Left one question regarding tests to get another opinion from @MaineC.

File: core/src/test/java/org/elasticsearch/index/query/SpanOrQueryBuilderTest.java (hunk 4 above)
Comment (cbuescher) [inline core/src/test/java/org/elasticsearch/index/query/SpanOrQueryBuilderTest.java]: I my last commit (after rebase) I added a utility method to SpanTermQueryBuilderTest that outputs an array of random SpanTermQueryBuilders with same fieldname and configurable size. Let me know what you think, I was able to reuse that in the other tests as well but maybe it's confusing? Let me know what you think.

**Comment by MaineC** [comment]
After checking the modifications in person with @cbuescher this morning: Looks all good to me.

**Pull request merged**